### ISSUE TITLE
objstor/aliyunoss: not upload promethues data to aliyunoss storage

### ISSUE THREAD

**Issue by kzcPo** [thanos-io/thanos#3198]
```
# prometheus --version
prometheus, version 2.20.1 (branch: HEAD, revision: 983ebb4a513302315a8117932ab832815f85e3d2)
  build user:       root@7cbd4d1c15e0
  build date:       20200805-17:26:58
  go version:       go1.14.6

# thanos --version
thanos, version 0.15.0 (branch: HEAD, revision: fbd14b49f95e7543883fedf9381a21127d4dda2b)
  build user:       root@d35b081d3e19
  build date:       20200907-09:29:21
  go version:       go1.14.2
```
**Object Storage Provider**:
ALIYUNOSS

**What happened**:
sidecar not upload promethues data to oss

**What you expected to happen**:
upload promethues data to oss

**prometheus startup parameter**:
```
prometheus \
--config.file=/etc/prometheus/prometheus-federate.yml \
--log.level=info \
--web.enable-admin-api \
--storage.tsdb.path=/data/prometheus-federate \
--web.listen-address=:9090 \
--web.external-url=http://0.0.0.0:9090/ \
--storage.tsdb.max-block-duration=2h \
--storage.tsdb.min-block-duration=2h \
--storage.tsdb.retention.time=30d \
--web.enable-lifecycle
```
**sidecar startup parameter**:
```
thanos sidecar \
--log.level=debug \
--prometheus.url=http://localhost:9090 \
--objstore.config-file=/etc/thanos/sidecar/oss.yaml \
--tsdb.path=/data/thanos/sidecar \
--shipper.upload-compacted
```
#### verify oss storage success
**thanos tools bucket verify --objstore.config-file=/etc/thanos/sidecar/oss.yaml**
```
level=info ts=2020-09-21T07:38:04.337313712Z caller=main.go:152 msg="Tracing will be disabled"
level=info ts=2020-09-21T07:38:04.337383155Z caller=factory.go:46 msg="loading bucket configuration"
level=warn ts=2020-09-21T07:38:04.341303018Z caller=verify.go:81 msg="GLOBAL COMPACTOR SHOULD __NOT__ BE RUNNING ON THE SAME BUCKET" issues=2 repair=false
level=info ts=2020-09-21T07:38:04.341333083Z caller=index_issue.go:33 msg="started verifying issue" with-repair=false issue=index_issue
level=info ts=2020-09-21T07:38:04.365387366Z caller=fetcher.go:453 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=24.032237ms cached=0 returned=0 partial=0
level=info ts=2020-09-21T07:38:04.365456323Z caller=index_issue.go:126 msg="verified issue" with-repair=false issue=index_issue
level=info ts=2020-09-21T07:38:04.365477448Z caller=overlapped_blocks.go:30 msg="started verifying issue" with-repair=false issue=overlapped_blocks
level=info ts=2020-09-21T07:38:04.370475205Z caller=fetcher.go:453 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=4.988509ms cached=0 returned=0 partial=0
level=info ts=2020-09-21T07:38:04.370498484Z caller=verify.go:100 msg="verify completed" issues=2 repair=false
level=info ts=2020-09-21T07:38:04.370652414Z caller=main.go:215 msg=exiting
```
**sidecar log**:
```
level=debug ts=2020-09-21T03:42:22.616629264Z caller=main.go:103 msg="maxprocs: Leaving GOMAXPROCS=[4]: CPU quota undefined"
level=info ts=2020-09-21T03:42:22.61681159Z caller=main.go:152 msg="Tracing will be disabled"
level=info ts=2020-09-21T03:42:22.616987253Z caller=options.go:23 protocol=gRPC msg="disabled TLS, key and cert must be set to enable"
level=info ts=2020-09-21T03:42:22.617054458Z caller=grpc.go:93 service=gRPC/server component=sidecar msg="registering as gRPC StoreAPI and RulesAPI"
level=info ts=2020-09-21T03:42:22.617277377Z caller=factory.go:46 msg="loading bucket configuration"
level=error ts=2020-09-21T03:42:22.621144402Z caller=sidecar.go:258 err="WAL dir is not accessible. Is this dir a TSDB directory? If yes it is shared with TSDB?: stat /data/thanos/sidecar/wal: no such file or directory"
level=info ts=2020-09-21T03:42:22.621188254Z caller=sidecar.go:299 msg="starting sidecar"
level=info ts=2020-09-21T03:42:22.621268774Z caller=reloader.go:222 component=reloader msg="started watching config file and recursively rule dirs for changes" cfg= out= dirs=
level=info ts=2020-09-21T03:42:22.621336949Z caller=intrumentation.go:48 msg="changing probe status" status=ready
level=info ts=2020-09-21T03:42:22.621577753Z caller=grpc.go:120 service=gRPC/server component=sidecar msg="listening for serving gRPC" address=0.0.0.0:10901
level=info ts=2020-09-21T03:42:22.621577936Z caller=intrumentation.go:60 msg="changing probe status" status=healthy
level=info ts=2020-09-21T03:42:22.621606749Z caller=http.go:57 service=http/server component=sidecar msg="listening for requests and metrics" address=0.0.0.0:10902
level=info ts=2020-09-21T03:42:22.624418574Z caller=sidecar.go:165 msg="successfully loaded prometheus external labels" external_labels="{region=\"wanli\", replica=\"1\"}"
level=info ts=2020-09-21T03:42:22.624444313Z caller=intrumentation.go:48 msg="changing probe status" status=ready
level=debug ts=2020-09-21T03:43:01.212643633Z caller=prometheus.go:259 msg="started handling ReadRequest_STREAMED_XOR_CHUNKS streamed read response."
level=debug ts=2020-09-21T03:43:01.212738543Z caller=prometheus.go:335 msg="handled ReadRequest_STREAMED_XOR_CHUNKS request." frames=0 series=0
level=debug ts=2020-09-21T03:44:27.640770306Z caller=prometheus.go:259 msg="started handling ReadRequest_STREAMED_XOR_CHUNKS streamed read response."
level=debug ts=2020-09-21T03:44:27.676643162Z caller=prometheus.go:335 msg="handled ReadRequest_STREAMED_XOR_CHUNKS request." frames=735 series=735
```

**prometheus log**:
```
level=info ts=2020-09-21T03:05:47.808Z caller=main.go:343 msg="Starting Prometheus" version="(version=2.20.1, branch=HEAD, revision=983ebb4a513302315a8117932ab832815f85e3d2)"
level=info ts=2020-09-21T03:05:47.808Z caller=main.go:344 build_context="(go=go1.14.6, user=root@7cbd4d1c15e0, date=20200805-17:26:58)"
level=info ts=2020-09-21T03:05:47.808Z caller=main.go:345 host_details="(Linux 3.10.0-957.21.3.el7.x86_64 #1 SMP Tue Jun 18 16:35:19 UTC 2019 x86_64 argo-prometheus-federate-H001 (none))"
level=info ts=2020-09-21T03:05:47.808Z caller=main.go:346 fd_limits="(soft=65535, hard=65535)"
level=info ts=2020-09-21T03:05:47.809Z caller=main.go:347 vm_limits="(soft=unlimited, hard=unlimited)"
level=info ts=2020-09-21T03:05:47.810Z caller=web.go:524 component=web msg="Start listening for connections" address=:9090
level=info ts=2020-09-21T03:05:47.810Z caller=main.go:684 msg="Starting TSDB ..."
level=info ts=2020-09-21T03:05:47.811Z caller=repair.go:59 component=tsdb msg="Found healthy block" mint=1600560000000 maxt=1600567200000 ulid=01EJMQ8TZSDE8QJ1P6VXD26V6S
level=info ts=2020-09-21T03:05:47.811Z caller=repair.go:59 component=tsdb msg="Found healthy block" mint=1600567200000 maxt=1600574400000 ulid=01EJMY4HZV9Q43RFWPWFJ3AR22
level=info ts=2020-09-21T03:05:47.811Z caller=repair.go:59 component=tsdb msg="Found healthy block" mint=1600574400000 maxt=1600581600000 ulid=01EJN509CJH6VCMXN22Q5KDAG7
level=info ts=2020-09-21T03:05:47.811Z caller=repair.go:59 component=tsdb msg="Found healthy block" mint=1600581600000 maxt=1600588800000 ulid=01EJNBW0N1CVT4YPQBTQQFHM0R
level=info ts=2020-09-21T03:05:47.811Z caller=repair.go:59 component=tsdb msg="Found healthy block" mint=1600588800000 maxt=1600596000000 ulid=01EJNJQQWEQA67CTZ56P5BY58P
level=info ts=2020-09-21T03:05:47.811Z caller=repair.go:59 component=tsdb msg="Found healthy block" mint=1600596000000 maxt=1600603200000 ulid=01EJNSKF324AEHMPSFXQKCACNW
level=info ts=2020-09-21T03:05:47.811Z caller=repair.go:59 component=tsdb msg="Found healthy block" mint=1600603200000 maxt=1600610400000 ulid=01EJP0F6ASDQYDYV8KZQDTCJ14
level=info ts=2020-09-21T03:05:47.811Z caller=repair.go:59 component=tsdb msg="Found healthy block" mint=1600610400000 maxt=1600617600000 ulid=01EJP7AXA33DPA3CHP056ZRYJR
level=info ts=2020-09-21T03:05:47.811Z caller=repair.go:59 component=tsdb msg="Found healthy block" mint=1600617600000 maxt=1600624800000 ulid=01EJPE6MSHCF9ZZYR3HAF70CMV
level=info ts=2020-09-21T03:05:47.811Z caller=repair.go:59 component=tsdb msg="Found healthy block" mint=1600624800000 maxt=1600632000000 ulid=01EJPN2CE767DJ0W0JE13BAATQ
level=info ts=2020-09-21T03:05:47.811Z caller=repair.go:59 component=tsdb msg="Found healthy block" mint=1600632000000 maxt=1600639200000 ulid=01EJPVY3795771HBQ5YETRFK6Y
level=info ts=2020-09-21T03:05:47.811Z caller=repair.go:59 component=tsdb msg="Found healthy block" mint=1600639200000 maxt=1600646400000 ulid=01EJQ2STJV9MQ1D8K21GC8MEHE
level=info ts=2020-09-21T03:05:47.811Z caller=repair.go:59 component=tsdb msg="Found healthy block" mint=1600646400000 maxt=1600653600000 ulid=01EJQ9NHXQVVJ97KD2CGVQJQQZ
level=info ts=2020-09-21T03:05:47.837Z caller=head.go:641 component=tsdb msg="Replaying on-disk memory mappable chunks if any"
level=info ts=2020-09-21T03:05:47.915Z caller=head.go:655 component=tsdb msg="On-disk memory mappable chunks replay completed" duration=78.050588ms
level=info ts=2020-09-21T03:05:47.915Z caller=head.go:661 component=tsdb msg="Replaying WAL, this may take a while"
level=info ts=2020-09-21T03:05:48.890Z caller=head.go:687 component=tsdb msg="WAL checkpoint loaded"
level=info ts=2020-09-21T03:05:50.057Z caller=head.go:713 component=tsdb msg="WAL segment loaded" segment=90 maxSegment=94
level=info ts=2020-09-21T03:05:50.468Z caller=head.go:713 component=tsdb msg="WAL segment loaded" segment=91 maxSegment=94
level=info ts=2020-09-21T03:05:50.657Z caller=head.go:713 component=tsdb msg="WAL segment loaded" segment=92 maxSegment=94
level=info ts=2020-09-21T03:05:50.772Z caller=head.go:713 component=tsdb msg="WAL segment loaded" segment=93 maxSegment=94
level=info ts=2020-09-21T03:05:50.772Z caller=head.go:713 component=tsdb msg="WAL segment loaded" segment=94 maxSegment=94
level=info ts=2020-09-21T03:05:50.772Z caller=head.go:716 component=tsdb msg="WAL replay completed" checkpoint_replay_duration=975.410367ms wal_replay_duration=1.881945797s total_replay_duration=2.935466109s
level=info ts=2020-09-21T03:05:51.006Z caller=main.go:700 fs_type=EXT4_SUPER_MAGIC
level=info ts=2020-09-21T03:05:51.006Z caller=main.go:701 msg="TSDB started"
level=info ts=2020-09-21T03:05:51.007Z caller=main.go:805 msg="Loading configuration file" filename=/etc/prometheus/prometheus-federate.yml
level=info ts=2020-09-21T03:05:51.007Z caller=main.go:833 msg="Completed loading of configuration file" filename=/etc/prometheus/prometheus-federate.yml
level=info ts=2020-09-21T03:05:51.007Z caller=main.go:652 msg="Server is ready to receive web requests."
level=info ts=2020-09-21T05:00:18.074Z caller=compact.go:495 component=tsdb msg="write block" mint=1600653600000 maxt=1600660800000 ulid=01EJQGH91KCNQM3T6HYMXA89VR duration=4.839051619s
level=info ts=2020-09-21T05:00:18.414Z caller=head.go:804 component=tsdb msg="Head GC completed" duration=124.66635ms
level=info ts=2020-09-21T05:00:18.449Z caller=checkpoint.go:96 component=tsdb msg="Creating checkpoint" from_segment=90 to_segment=92 mint=1600660800000
level=info ts=2020-09-21T05:00:20.416Z caller=head.go:884 component=tsdb msg="WAL checkpoint complete" first=90 last=92 duration=1.967459605s
level=info ts=2020-09-21T07:00:17.535Z caller=compact.go:495 component=tsdb msg="write block" mint=1600660800000 maxt=1600668000000 ulid=01EJQQD06BFJ5HDM9CK5BZPBT8 duration=4.403700233s
level=info ts=2020-09-21T07:00:17.875Z caller=head.go:804 component=tsdb msg="Head GC completed" duration=123.459228ms
level=info ts=2020-09-21T07:00:17.903Z caller=checkpoint.go:96 component=tsdb msg="Creating checkpoint" from_segment=93 to_segment=95 mint=1600668000000
level=info ts=2020-09-21T07:00:19.824Z caller=head.go:884 component=tsdb msg="WAL checkpoint complete" first=93 last=95 duration=1.921233402s
```
** ls sidecar data**:
```
# ls -l /data/thanos/sidecar/
total 8
-rw-r--r-- 1 root root   37 Sep 21 15:43 thanos.shipper.json
drwxr-xr-x 2 root root 4096 Sep 21 11:42 wal
```
**more  thanos.shipper.json**:
```
{
	"version": 1,
	"uploaded": null
}
```
** ls prometheus data**:
```
total 72
drwxr-xr-x 3 root root  4096 Sep 20 11:00 01EJMQ8TZSDE8QJ1P6VXD26V6S
drwxr-xr-x 3 root root  4096 Sep 20 13:00 01EJMY4HZV9Q43RFWPWFJ3AR22
drwxr-xr-x 3 root root  4096 Sep 20 15:00 01EJN509CJH6VCMXN22Q5KDAG7
drwxr-xr-x 3 root root  4096 Sep 20 17:00 01EJNBW0N1CVT4YPQBTQQFHM0R
drwxr-xr-x 3 root root  4096 Sep 20 19:00 01EJNJQQWEQA67CTZ56P5BY58P
drwxr-xr-x 3 root root  4096 Sep 20 21:00 01EJNSKF324AEHMPSFXQKCACNW
drwxr-xr-x 3 root root  4096 Sep 20 23:00 01EJP0F6ASDQYDYV8KZQDTCJ14
drwxr-xr-x 3 root root  4096 Sep 21 01:00 01EJP7AXA33DPA3CHP056ZRYJR
drwxr-xr-x 3 root root  4096 Sep 21 03:00 01EJPE6MSHCF9ZZYR3HAF70CMV
drwxr-xr-x 3 root root  4096 Sep 21 05:00 01EJPN2CE767DJ0W0JE13BAATQ
drwxr-xr-x 3 root root  4096 Sep 21 07:00 01EJPVY3795771HBQ5YETRFK6Y
drwxr-xr-x 3 root root  4096 Sep 21 09:00 01EJQ2STJV9MQ1D8K21GC8MEHE
drwxr-xr-x 3 root root  4096 Sep 21 11:00 01EJQ9NHXQVVJ97KD2CGVQJQQZ
drwxr-xr-x 3 root root  4096 Sep 21 13:00 01EJQGH91KCNQM3T6HYMXA89VR
drwxr-xr-x 3 root root  4096 Sep 21 15:00 01EJQQD06BFJ5HDM9CK5BZPBT8
drwxr-xr-x 2 root root  4096 Sep 21 15:00 chunks_head
-rw-r--r-- 1 root root     0 Sep 17 18:06 lock
-rw-r--r-- 1 root root 20001 Sep 21 11:05 queries.active
drwxr-xr-x 3 root root  4096 Sep 21 15:00 wal
```

**Comment by bwplotka** [comment]
Hey

Thanks for reporting & all logs.

I think I found a reason of the behaviour you see. From sidecar logs:

> level=error ts=2020-09-21T03:42:22.621144402Z caller=sidecar.go:258 err="WAL dir is not accessible. Is this dir a TSDB directory? If yes it is shared with TSDB?: stat /data/thanos/sidecar/wal: no such file or directory"

This means that you maybe made some mistake in sidecar flag to tsdb path, so sidecar does have access to Prometheus dir (: Can you double check that? I seen it couple of times already, it's easy to miss, that's why error. I think we should even crash sidecar on it, WDYT?

**Comment by kzcPo** [comment]
@bwplotka Hello!
```
/data/thanos/sidecar/wal
```
This folder will not be created automatically. Manual creation solves this error.
It is suggested that the TSDB directory can be automatically created in the future.
However, this problem still exists and cannot be uploaded to aliyunoss bucket.
I'm sure I have access to TSDB.

Is there any other way to force the upload action or provide more error information?

**Comment by tanujd11** [comment]
@bwplotka @kzcPo Getting the same error but  the sidecar coming to ready state. Will this work? If not, could you help?

Here are the sidecar logs

```
level=info ts=2020-11-04T08:02:47.241039059Z caller=main.go:152 msg="Tracing will be disabled"
level=info ts=2020-11-04T08:02:47.241243965Z caller=options.go:23 protocol=gRPC msg="disabled TLS, key and cert must be set to enable"
level=info ts=2020-11-04T08:02:47.241321856Z caller=grpc.go:93 service=gRPC/server component=sidecar msg="registering as gRPC StoreAPI and RulesAPI"
level=info ts=2020-11-04T08:02:47.241574941Z caller=factory.go:46 msg="loading bucket configuration"
level=error ts=2020-11-04T08:02:47.241890026Z caller=sidecar.go:258 err="WAL dir is not accessible. Is this dir a TSDB directory? If yes it is shared with TSDB?: stat /prometheus/wal: no such file or directory"
level=info ts=2020-11-04T08:02:47.24192366Z caller=sidecar.go:299 msg="starting sidecar"
level=info ts=2020-11-04T08:02:47.242108203Z caller=reloader.go:222 component=reloader msg="started watching config file and recursively rule dirs for changes" cfg= out= dirs=
level=info ts=2020-11-04T08:02:47.242147661Z caller=intrumentation.go:60 msg="changing probe status" status=healthy
level=info ts=2020-11-04T08:02:47.242215825Z caller=http.go:57 service=http/server component=sidecar msg="listening for requests and metrics" address=[10.136.38.10]:10902
level=info ts=2020-11-04T08:02:47.242419066Z caller=intrumentation.go:48 msg="changing probe status" status=ready
level=info ts=2020-11-04T08:02:47.24246633Z caller=grpc.go:120 service=gRPC/server component=sidecar msg="listening for serving gRPC" address=[10.136.38.10]:10901
level=warn ts=2020-11-04T08:02:47.242883191Z caller=sidecar.go:311 msg="failed to get Prometheus flags. Is Prometheus running? Retrying" err="request config against http://localhost:9090/api/v1/status/flags: Get \"http://localhost:9090/api/v1/status/flags\": dial tcp 127.0.0.1:9090: connect: connection refused"
level=info ts=2020-11-04T08:02:49.265089589Z caller=sidecar.go:165 msg="successfully loaded prometheus external labels" external_labels="{prometheus=\"monitoring/k8s\", prometheus_replica=\"prometheus-k8s-0\"}"
level=info ts=2020-11-04T08:02:49.26513144Z caller=intrumentation.go:48 msg="changing probe status" status=ready
```

**Comment by xiaodangjia457** [comment]
hello i have encountered the same problem that can't upload the Prometheus data to OSS, the error info: 

upload debug meta file: failed to get size apriori to upload debug/metas/01EQS2******1Y5EQE3M.json: unsupported type of io.Reader" uploaded=0

do you have some solutions to fix this?

**Comment by xiaodangjia457** [comment]
@boosty

**Comment by xiaodangjia457** [comment]
@boosty  
hello i have encountered the same problem that can't upload the Prometheus data to OSS, the error info:

upload debug meta file: failed to get size apriori to upload debug/metas/01EQS2******1Y5EQE3M.json: unsupported type of io.Reader" uploaded=0

do you have some solutions to fix this?

**Comment by BegoniaGit** [comment]
I upgraded to ThanOS-v0.17.0 and had the same problem

**Comment by xiaodangjia457** [comment]
> I upgraded to Thanos-v0.17.0 and had the same problem

use the 0.16.0 will be fine

**Comment by chris-ng-scmp** [comment]
yes, the same issue with https://github.com/thanos-io/thanos/issues/3302

only working back to 0.16.0

**Comment by chris-ng-scmp** [comment]
Compactor having the same issue in 0.17

```
level=error ts=2021-01-21T03:22:59.430377095Z caller=compact.go:434 msg="retriable error" err="compaction: group 0@17005465453309916553: upload of 01EWHF6PTZ22P35GVVE9XEMHPE failed: upload debug meta file: failed to get size apriori to upload debug/metas/01EWHF6PTZ22P35GVVE9XEMHPE.json: unsupported type of io.Reader"
```

**Comment by yeya24** [comment]
@xiaodangjia457  @chris-ng-scmp 
The unsupported reader problem has been fixed by https://github.com/thanos-io/thanos/pull/3476. Please try the latest release.

**Comment by heruscoder** [comment]
> @xiaodangjia457 @chris-ng-scmp
> The unsupported reader problem has been fixed by #3476. Please try the latest release.

I think the latest release doesnt cover this fix yet
